File: lib/agent/nodes/responder.ts
import { AIMessage } from '@langchain/core/messages';
import type { AgentStateType } from '../state';

export function responderNode(
  state: AgentStateType
): Partial<AgentStateType> {
  // If there's an error and no finalResponse already set, surface it
  if (state.error && !state.finalResponse) {
    const errMsg = state.error;
    const aiMessage = new AIMessage(`Error: ${errMsg}`);
    return {
      messages: [aiMessage],
      finalResponse: {
        type: 'error',
        message: errMsg,
      },
    };
  }

  if (state.intent === 'greeting') {
    const msg = "Hi! I'm your SQL assistant. Ask me to query your database, explore your tables, or modify data.";
    const aiMessage = new AIMessage(msg);
    return {
      messages: [aiMessage],
      finalResponse: {
        type: 'greeting',
        message: msg,
      },
    };
  }

  if (state.intent === 'out_of_scope') {
    const msg = "I'm focused on database queries. I can help you fetch, analyze, or modify data in your connected database.";
    const aiMessage = new AIMessage(msg);
    return {
      messages: [aiMessage],
      finalResponse: {
        type: 'answer',
        message: msg,
      },
    };
  }

  if (state.intent === 'table_structure') {
    const msg = state.schemaFormatted || 'No schema information available.';
    const aiMessage = new AIMessage(`Table Structure:\n${msg}`);
    return {
      messages: [aiMessage],
      finalResponse: {
        type: 'schema_info',
        message: msg,
      },
    };
  }

  // Fallback — if finalResponse was already set by an earlier node, pass through
  if (state.finalResponse) {
    const msgText = state.finalResponse.message || state.finalResponse.explanation || '';
    const aiMessage = new AIMessage(msgText);
    return {
      messages: [aiMessage],
    };
  }

  const defaultMsg = 'Something unexpected happened. Please try again.';
  return {
    messages: [new AIMessage(defaultMsg)],
    finalResponse: {
      type: 'error',
      message: defaultMsg,
    },
  };
}

File: backend/agent/nodes/responder.py
from typing import Dict, Any
from langchain_core.messages import AIMessage

def responder_node(state: Dict[str, Any]) -> Dict[str, Any]:
    error = state.get("error")
    final_resp = state.get("final_response")
    intent = state.get("intent")

    # If there's an error and no finalResponse already set, surface it
    if error and not final_resp:
        ai_message = AIMessage(content=f"Error: {error}")
        return {
            "messages": [ai_message],
            "final_response": {
                "type": "error",
                "message": error
            }
        }

    if intent == 'greeting':
        msg = "Hi! I'm your SQL assistant. Ask me to query your database, explore your tables, or modify data."
        ai_message = AIMessage(content=msg)
        return {
            "messages": [ai_message],
            "final_response": {
                "type": "greeting",
                "message": msg
            }
        }

    if intent == 'out_of_scope':
        msg = "I'm focused on database queries. I can help you fetch, analyze, or modify data in your connected database."
        ai_message = AIMessage(content=msg)
        return {
            "messages": [ai_message],
            "final_response": {
                "type": "answer",
                "message": msg
            }
        }

    if intent == 'table_structure':
        msg = state.get("schema_formatted") or "No schema information available."
        ai_message = AIMessage(content=f"Table Structure:\n{msg}")
        return {
            "messages": [ai_message],
            "final_response": {
                "type": "schema_info",
                "message": msg
            }
        }

    # Fallback — if finalResponse was already set by an earlier node, pass through
    if final_resp:
        msg_text = final_resp.get("message") or final_resp.get("explanation") or ""
        ai_message = AIMessage(content=msg_text)
        return {
            "messages": [ai_message]
        }

    default_msg = "Something unexpected happened. Please try again."
    return {
        "messages": [AIMessage(content=default_msg)],
        "final_response": {
            "type": "error",
            "message": default_msg
        }
    }
